Implement Shoot (Stage B), resolves #7

HN67 opened the pull request:
Implemented Projectiles (which have a finite lifetime) and shooting, using command "shoot". The code works, but two issues are becoming more apparent. They are probably fine for now but we might want to refactor the code before too long (maybe once we create the new interfaces, Loot and Rect especially).
1) Control mechanism: I struggled a bit with deciding how to decide when a Player shoots, and how to pass the new Projectile back up to the Room. I ended up implemented `boolean shoots()` and `Projectile shoot()`, and then Room gets a projectile from shoot() if shoots() is true. I do like this method, but we should probably decide on a formal architecture.
2) Class inheritance/composition: I think ideally the shoot/shoots() methods would belong to a Shooter interface, and then we can abstract away from Player a bit. Maybe even compose a Shooter into Player and have Player implement a Shootable interface. Also, attributes like lifetime would be better if they were not locked to Projectile, maybe in a Mortal interface or something.

HN67 committed:
Move update movement to DefaultEntity, add alive() method to specification

HN67 committed:
Create Projectile entity, which is only alive when it has a positive age

Age decreases every update

HN67 committed:
Make a Room clear Entities that are no longer alive

Also added testing projectile

HN67 committed:
Upgrade Player to be able to create a useful Projectile

HN67 committed:
Add shoots method to Player to allow higher control of shooting

HN67 committed:
Change Room specification to specify adding general entity/player, change textroom to handle player shooting

rupertraphael commented on src/Player.java:
```diff
@@ -3,30 +3,73 @@
 public class Player extends DefaultEntity {
 
+    // Temporary variable, should be moved to config or something?
+    private int bulletLife = 5;
```
I was thinking that we might put this in a concrete Projectile class like, let's say, Missile? Life (or in my words, range) could be defined by every concrete Projectile class that we have. Not by the player.

rupertraphael replied:
Of course, we might expose this instance variable via a setter just in case Effect might manipulate a Projectile's life.

rupertraphael replied:
For example, an ExtendsRange Effect might call missile.setLifetime(5). And maybe I'm just trying to get ahead of myself but I couldn't help but think about also having a SpeedUp Effect that might call missile.setVelocity(someVector).

rupertraphael replied:
But I guess for now, I'd take this off from Player and just set a default Projectile lifetime for the sake of MVP.

rupertraphael commented on src/TextRoom.java:
```diff
@@ -6,20 +6,41 @@
     }
 
-    public void add(Entity entity) {
+    public void addEntity(Entity entity) {
```
Much more a preference rather than a needed change; I'd like to call this method _spawn_.

rupertraphael reviewed (commented):
@HN67, I commented on your question about where to put the bulletLife and added a preference regarding your addEntity method under TextRoom. As what I've said in Discord, I might do some more reviewing after my midterm.

rupertraphael commented:
> Control mechanism: I struggled a bit with deciding how to decide when a Player shoots, and how to pass the new Projectile back up to the Room. I ended up implemented boolean shoots() and Projectile shoot(), and then Room gets a projectile from shoot() if shoots() is true. I do like this method, but we should probably decide on a formal architecture.

Similar to move, I'd like to think of shoot as something that's invoked by Player. I guess, what I'm trying to say here is that in the Controller, inside the while loop, we could call a `player.act(command, room)` method. The act method should decide whether to move the player or let it shoot a projectile depending on the command string.

For me, this is kinda intuitive because the Controller is controlling the player via the act method that I just mentioned. Also, I think one advantage is that we won't be invoking each of the existing entities' update methods when the user inputs a command. Of course, this becomes a disadvantage if we want this for a specific reason.

HN67 committed:
Rename Room add methods to spawn methods

HN67 committed:
Make Player spawn a Bullet, which manages its own lifetime

HN67 committed:
Abstract Player shooting abilities into Spawner interface

HN67 committed:
Fix bug in Player movement where velocity persists after movement command

HN67 commented:
I have introduced a concrete Bullet class which has a fixed lifetime, and then Player spawns that, instead of managing lifetime itself. I have refactored Room methods to replace 'add' with 'spawn', and have also abstracted the Player's shooting abilities into a Spawner interface, which is then referenced in the Room's update method.

> Similar to move, I'd like to think of shoot as something that's invoked by Player. I guess, what I'm trying to say here is that in the Controller, inside the while loop, we could call a `player.act(command, room)` method. The act method should decide whether to move the player or let it shoot a projectile depending on the command string.

I have some reservations about using the `.act` architecture. One of my design goals is to decouple the classes as much as possible, so that each class has as few dependencies as possible, since every dependency a class has means that if the dependency is changed, the class also likely will need to be updated to deal with the changes. This means two main things: first, ideally a class does not even know about most other classes, e.g. right now Entity (and Player) do not touch Room at all; second, when a class does touch another, it should interact in a minimal fashion, e.g. using public methods like .update (or .act) and letting the other class handle the logic. Obviously these two goals are in slight conflict, since the less Player knows the more Room has to handle itself. Additionally, the more direct the class hierarchy and dependency tree is, the better.

Using the Spawner architecture (what I have set up currently) creates a dependency tree of Controller -> Room -> Player. The Controller only knows about a Room, and the Room only knows about Entity/Player, and Player doesn't know about either of them. In total, this is two dependencies, and a linear dependency tree.

Using the Actor (.act) architecture creates a dependency tree of Controller -> Room & Player, Room -> Entity/Player, and Player -> Room. In total, this is four dependencies, adding a new one to Controller and a new one to Player, which previously had none. Additionally, the dependency tree is now a diamond shape, with a loop attached between Player and Room, tightly coupling them.

This is why I believe the Spawner architecture to be more flexible and extensible, since it minimizes the dependency complexity. Additionally, the Spawner architecture does not add considerable logic to Room, since Spawners provide their own .spawns and .spawn methods.

rupertraphael committed:
Fix bug where player velocity was always set to 0 when not going right.

rupertraphael committed:
Implement projectile speed.

rupertraphael committed:
Remove extra print statement.

rupertraphael reviewed (approved):
@HN67 you may merge unless you have more changes to add.
